pkg/controller/petset/pet_set.go

```diff
@@ -0,0 +1,342 @@
+	ps, err := psc.psStore.GetPodPetSets(pod)
+	if err != nil {
+		glog.V(4).Infof("No PetSets found for pod %v, PetSet controller will avoid syncing", pod.Name)
+		return nil
+	}
+	// Resolve a overlapping petset tie by creation timestamp.
+	// Let's hope users don't create overlapping petsets.
+	if len(ps) > 1 {
+		glog.Errorf("user error! more than one PetSet is selecting pods with labels: %+v", pod.Labels)
+		sort.Sort(overlappingPetSets(ps))
+	}
+	return &ps[0]
+}
+
+func (psc *PetSetController) enqueuePetSet(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Cound't get key for object %+v: %v", obj, err)
+		return
+	}
+	psc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (psc *PetSetController) worker() {
+	for {
+		func() {
+			key, quit := psc.queue.Get()
+			if quit {
+				return
+			}
+			defer psc.queue.Done(key)
+			if errs := psc.syncHandler(key.(string)); len(errs) != 0 {
+				glog.Errorf("Error syncing PetSet, requeuing: %v", errs)
+				psc.queue.Add(key)
+			}
+		}()
+	}
+}
+
+// Sync syncs the given petset.
+func (psc *PetSetController) Sync(key string) []error {
+	startTime := time.Now()
+	defer func() {
+		glog.Infof("Finished syncing pet set %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	if !psc.podStoreSynced() {
+		// Sleep so we give the pod reflector goroutine a chance to run.
+		time.Sleep(PodStoreSyncedPollPeriod)
+		return []error{fmt.Errorf("Waiting for pods controller to sync")}
+	}
+
+	obj, exists, err := psc.psStore.Store.GetByKey(key)
+	if !exists {
+		glog.Infof("PetSet has been deleted %v", key)
+		return []error{}
+	}
+	if err != nil {
+		glog.Infof("Unable to retrieve PetSet %v from store: %v", key, err)
+		return []error{err}
+	}
+
+	ps := *obj.(*apps.PetSet)
+	petList, err := psc.getPodsForPetSet(&ps)
+	if err != nil {
+		return []error{err}
+	}
+
+	numPets, errs := psc.syncPetSet(&ps, petList)
+	if err := updatePetCount(psc.kubeClient, ps, numPets); err != nil {
+		glog.V(2).Infof("Failed to update replica count for petset %v/%v; requeuing; error: %v", ps.Namespace, ps.Name, err)
+		errs = append(errs, err)
+	}
+
+	return errs
+}
+
+// syncPetSet syncs a tuple of (petset, pets).
+func (psc *PetSetController) syncPetSet(ps *apps.PetSet, pets []*api.Pod) (int, []error) {
+	glog.Infof("Syncing PetSet %v/%v with %d pets", ps.Namespace, ps.Name, len(pets))
+
+	it := NewPetSetIterator(ps, pets)
+	blockingPet, err := psc.blockingPetStore.Get(ps, pets)
+	if err != nil {
+		return 0, []error{err}
+	}
+	if blockingPet != nil {
+		glog.Infof("PetSet %v blocked from scaling on pet %v", ps.Name, blockingPet.pod.Name)
+	}
+	petManager := psc.newSyncer(blockingPet)
+	numPets := 0
+
+	for it.Next() {
+		pet := it.Value()
+		if pet == nil {
+			continue
+		}
+		switch pet.event {
+		case syncPet:
+			err = petManager.Sync(pet)
```

Comment:
If I get a "sync" event here, it's not clear what to do here.
Can you provide some details about what's going on?